### PR TITLE
Fold all local addresses in local morph

### PR DESCRIPTION
**Pull request by SingleAccretion**
This change enables folding for all local address trees in local morph. This is a requirement for the next round of simplifications enabled by the `ADDR` work (e. g. #78246 depends on this).

There is some amount of (mostly positive) [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=101392&view=ms.vss-build-web.run-extensions-tab) due to more forward substitutions enabled by the fact `LCL_ADDR` nodes by themselves do not need `GLOB_REF`. Some regressions in tests due to a bit more `GLOB_REF`ing for "wide" indirs. One case in a huge test method where swapping `LCL_VAR` and `LCL_VAR_ADDR` operands for a relop leads us to miss out on a lot of `mov` suppression.

There is also a very nice TP win, up to `1%` reduction in instructions retired.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib 

@jakobbotsch This should solve the problem of return buffer calls not having defined address shape before morph.

**Comment by BruceForstall** [comment]
fyi, I reran the diffs/replay pipelines because they failed due to missing osx-arm64 MCH files. There should be some there now (but we're still missing libraries-pmi, which failed during collection)

File: src/coreclr/jit/flowgraph.cpp [hunk 1]
```diff
@@ -754,29 +754,6 @@ GenTreeLclVar* Compiler::fgIsIndirOfAddrOfLocal(GenTree* tree)
...
-            {
-                if (base->OperGet() == GT_IND)
-                {
-                    return fgIsIndirOfAddrOfLocal(base);
-                }
-                // else use base as addr
-                addr = base;
-            }
-        }
-
```
Comment (jakobbotsch) [inline src/coreclr/jit/flowgraph.cpp:779]: Couldn't we still have this (maybe not for tracked locals)?

Comment (jakobbotsch) [reply src/coreclr/jit/flowgraph.cpp:779]: Right, I mean for untracked locals. The name of the function (`fgIsIndirOfAddrOfLocal`) does not seem to suggest that this function should only work for tracked locals, but I guess we don't really use it for any other purpose.

Comment (jakobbotsch) [reply src/coreclr/jit/flowgraph.cpp:779]: Can we unify some of these functions now? `fgIsIndirOfAddrOfLocal`, `GenTree::IsLocalAddrExpr`, `GenTree::DefinesLocalAddr`, `impIsAddressInLocal`? I guess the latter one not, but maybe some of the others? (Not necessarily in this PR)

**Comment by jakobbotsch** [comment]
Looks great to me! I will run some stress jobs.

File: src/coreclr/jit/optimizer.cpp [hunk 2]
```diff
@@ -6021,27 +6021,37 @@ bool Compiler::optIsVarAssignedWithDesc(Statement* stmt, isVarAssgDsc* dsc)
...
             {
                 m_dsc->ivaMaskCall = optCallInterf(tree->AsCall());
-
-                dest = m_compiler->gtCallGetDefinedRetBufLclAddr(tree->AsCall());
-                if (dest == nullptr)
-                {
-                    return WALK_CONTINUE;
-                }
-
-                dest = dest->AsOp()->gtOp1;
```
[the remark this answers is not in the document]
Comment (SingleAccretion) [reply src/coreclr/jit/optimizer.cpp:6037]: It is still needed and should be handled below by `DefinesLocal`.

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**